core/database_util.go

```diff
@@ -164,6 +95,20 @@ func GetHeadFastBlockHash(db ethdb.Database) common.Hash {
 	return common.BytesToHash(data)
 }
 
+// GetHeadBlockNum retrieves the block number of the current canonical head block.
+func GetHeadBlockNum(db ethdb.Database) *big.Int {
```

Comment:
All in all, this method is not warranted for. It just combines multiple (already implemented) database ops into one, resulting in unnecessary API bloat and a potential for added bugs (which is spectacularly demonstrate). I'm all for deleting this.